### PR TITLE
KAFKA-5172: Fix fetchPrevious to find the correct session.

### PR DESCRIPTION
**Pull request by ghost**
Change fetchPrevious to use findSessions with the proper key and timestamps rather than using fetch.

### CODE REVIEW

**Comment by guozhangwang** [comment]
@dguy could you take a look?

**Review by dguy — COMMENTED** [review]
@KyleWinkelman thanks for the PR. Can you please add a test to `CachingSessionStoreTest` that fails without this change?

**Comment by ghost** [comment]
@dguy I have added a test case that fails without this change.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java [hunk 1]
```diff
@@ -16,35 +16,37 @@
  */
 package org.apache.kafka.streams.state.internals;
 
+import static org.apache.kafka.streams.state.internals.RocksDBSessionStoreTest.toList;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java]: can you re-order the input back to what they where before?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java [hunk 2]
```diff
@@ -162,6 +164,22 @@ public void shouldFetchCorrectlyAcrossSegments() throws Exception {
     }
 
     @Test
+    public void shouldPutAndMaybeForward() throws Exception {
+        final Windowed<String> a1 = new Windowed<>("a", new SessionWindow(0, 0));
+        final Windowed<String> a2 = new Windowed<>("a", new SessionWindow(0, 1));
+        final MockCacheFlushListener<Windowed<String>, Long> flushListener = new MockCacheFlushListener<>();
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java]: I don't think we need to add the extra class just for this case. We could do something like:
```
final List<Change<Long>> values = new ArrayList<>();
        cachingStore.setFlushListener(new CacheFlushListener<Windowed<String>, Long>() {
            @Override
            public void apply(final Windowed<String> key, final Long newValue, final Long oldValue) {
                values.add(new Change<>(newValue, oldValue));
            }
        });
       ...
      assertThat(values, equalTo(Arrays.asList(new Change<>(1L, null), new Change<>(null, 1L), new Change<>(2L, null))));
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java [hunk 3]
```diff
@@ -162,6 +164,22 @@ public void shouldFetchCorrectlyAcrossSegments() throws Exception {
     }
 
     @Test
+    public void shouldPutAndMaybeForward() throws Exception {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java]: maybe something like: `shouldForwardChangedValuesDuringFlush` ?

File: streams/src/test/java/org/apache/kafka/test/MockCacheFlushListener.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import static org.junit.Assert.assertEquals;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.internals.CacheFlushListener;
+import org.apache.kafka.streams.kstream.internals.Change;
+
+public class MockCacheFlushListener<K, V> implements CacheFlushListener<K, V> {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/test/MockCacheFlushListener.java]: See comment above, i don't think we need this class

**Review by dguy — COMMENTED** [review]
Thanks for adding the test @KyleWinkelman left a couple of comments

**Comment by ghost** [comment]
I made the recommended changes and altered the contents of the test a little so we could see a case where the new and old values are set.
Thanks for the comments.

**Comment by ghost** [comment]
@dguy  Do you think we should prevent the Bytes.wrap(serdes.rawKey(key.key())) from being called twice (once in putAndMaybeForward and once in fetchPrevious)? I don't know how costly this is but it could save a little bit of processing. I think we could call it once and change the fetchPrevious to accept Bytes and Window.

```
    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
        final Bytes binaryKey = entry.key();
        final RecordContext current = context.recordContext();
        context.setRecordContext(entry.recordContext());
        try {
            final Windowed<K> key = SessionKeySerde.from(binaryKey.get(), keySerde.deserializer(), topic);
            **final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));**
            if (flushListener != null) {
                final AGG newValue = serdes.valueFrom(entry.newValue());
                final AGG oldValue = fetchPrevious(rawKey, key.window());
                if (!(newValue == null && oldValue == null)) {
                    flushListener.apply(key, newValue, oldValue);
                }
            }
            bytesStore.put(new Windowed<>(**rawKey**, key.window()), entry.newValue());
        } finally {
            context.setRecordContext(current);
        }
    }
```

```
    private AGG fetchPrevious(**final Bytes rawKey, final Window window**) {
        try (final KeyValueIterator<Windowed<Bytes>, byte[]> iterator = bytesStore
                .findSessions(**rawKey**, window.start(), window.end())) {
            if (!iterator.hasNext()) {
                return null;
            }
            return serdes.valueFrom(iterator.next().value);
        }
    }
```

**Comment by dguy** [comment]
@KyleWinkelman yes that makes sense. Thanks!

**Comment by ghost** [comment]
Ok made the changes. Anything else required of me for this PR?

**Review by dguy — APPROVED** [review]
Thanks @KyleWinkelman LGTM

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Review by guozhangwang — APPROVED** [review]
LGTM! Merged to trunk.

**Pull request closed without merging**